### ISSUE TITLE
Generic types not set when yielding from a contextlib.contextmanager

### ISSUE THREAD

**Issue by Terrance** [microsoft/pyright#1982]
## Describe the bug

When defining a context manager function using `contextlib.contextmanager`, which yields an object of generic type based on the function input, the type doesn't reach the produced value when used.

The function's docstring and argument type information are also lost, which is presumably why the input type isn't propagating to the output.

## To reproduce

```python
from contextlib import contextmanager
from typing import Generic, Iterator, TypeVar

T = TypeVar("T")

class Box(Generic[T]):
    def __init__(self, value: T):
        self.value = value

@contextmanager
def check_box(box: Box[T]) -> Iterator[T]:
    yield box.value

reveal_type(check_box)  # Type of "check_box" is "(*args: Any, **kwargs: Any) ->
                        # _GeneratorContextManager[T@check_box]

with check_box(Box(3)) as number:
    reveal_type(number)  # Type of "number" is "T@check_box"
```

## Expected behaviour

`number` here would ideally be an `int`.

## VS Code extension or command-line

Pyright 1.1.148 in VSCode.

**Comment by erictraut** [comment]
This is a problem with the type declaration of `contextmanager` in typeshed. It doesn't preserve the parameter types of the function it is decorating. Mypy has special-case knowledge of this function to work around this limitation. The correct fix is to use PEP 612 (`ParamSpec`) to properly annotate `contextmanager`. Unfortunately, we're not yet able to `ParamSpec` in typeshed because some type checkers (including mypy) don't yet support it. Refer to [this tracking bug](https://github.com/python/typeshed/issues/4827) for details. Once this bug is resolved, we'll be able to update the declaration for `contextmanager`, and this case should work as expected.

**Comment by jakebailey** [comment]
FWIW, typeshed was updated this morning to start allowing `ParamSpec` in a few situations, and the next time we pull from typeshed this should be fixed.

**Comment by erictraut** [comment]
I just pulled the latest version of the typeshed stubs into pyright and confirmed that the above code now works as expected.

**Comment by erictraut** [comment]
This is included in pyright 1.1.149, which I just published. It will also be in the next release of pylance.
